processor/otel/metadata.go

```diff
@@ -0,0 +1,130 @@
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package otel
+
+import (
+	"fmt"
+	"net"
+	"strings"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/beats/v7/libbeat/common"
+)
+
+func translateResourceMetadata(resource pdata.Resource, out *model.Metadata) {
+	var exporterVersion string
+	resource.Attributes().ForEach(func(k string, v pdata.AttributeValue) {
+		switch k {
+		case conventions.AttributeServiceName:
+			out.Service.Name = truncate(v.StringVal())
+		case conventions.AttributeServiceVersion:
+			out.Service.Version = truncate(v.StringVal())
+		case conventions.AttributeServiceInstance:
+			out.Service.Node.Name = truncate(v.StringVal())
+		case conventions.AttributeDeploymentEnvironment:
+			out.Service.Environment = truncate(v.StringVal())
+
+		case conventions.AttributeTelemetrySDKName:
+			out.Service.Agent.Name = truncate(v.StringVal())
+		case conventions.AttributeTelemetrySDKLanguage:
+			out.Service.Language.Name = truncate(v.StringVal())
+		case conventions.AttributeTelemetrySDKVersion:
+			out.Service.Agent.Version = truncate(v.StringVal())
+
+		case conventions.AttributeK8sNamespace:
+			out.System.Kubernetes.Namespace = truncate(v.StringVal())
+		case conventions.AttributeK8sPod:
+			out.System.Kubernetes.PodName = truncate(v.StringVal())
+		case conventions.AttributeK8sPodUID:
+			out.System.Kubernetes.PodUID = truncate(v.StringVal())
+
+		case conventions.AttributeHostName:
+			out.System.DetectedHostname = truncate(v.StringVal())
+
+		case conventions.AttributeProcessID:
+			out.Process.Pid = int(v.IntVal())
+
+		case conventions.OCAttributeExporterVersion:
+			exporterVersion = v.StringVal()
+
+		default:
+			if out.Labels == nil {
+				out.Labels = make(common.MapStr)
+			}
+			out.Labels[replaceDots(k)] = ifaceAttributeValue(v)
+		}
+	})
+
+	if strings.HasPrefix(exporterVersion, "Jaeger") {
+		// version is of format `Jaeger-<agentlanguage>-<version>`, e.g. `Jaeger-Go-2.20.0`
+		const nVersionParts = 3
+		versionParts := strings.SplitN(exporterVersion, "-", nVersionParts)
+		if out.Service.Language.Name == "" && len(versionParts) == nVersionParts {
+			out.Service.Language.Name = versionParts[1]
+		}
+		if v := versionParts[len(versionParts)-1]; v != "" {
```

Comment:
should this be `len(nVersionParts)-1` ?

Reply:
Nope.
 - `nVersionParts` is a constant (did you mean `nVersionParts-1?), and I believe this code is intended to cater for both `Jaeger-<version>` and `Jaeger-<language>-<version>`
 - This is existing code, just moved from consumer.go

Reply:
yes I actually meant `nVersionParts-1`; 
theoretically if `exporterVersion` is `Jaeger` we would set the `out.Service.Agent.Version=Jaeger` - but since it is existing code, I assume that is theoretically and ok to keep as is.